### ISSUE TITLE
Port the dreamerv3-torch Output to ROS2 Actions

### ISSUE THREAD

**Issue by waterkimchi** [uci-f1tenth/uci_f1tenth_workshop#9]
## Description

**PLEASE WORK ON THE DESIGNATED BRANCH, DO NOT CREATE ONE - port-dreamer-output**

Adapt DreamerV3 to control the F1Tenth car using steering and speed as the two action outputs. This involves configuring the action space of the DreamerV3 agent, ensuring proper action sampling, and implementing a robust mapping of the agent's actions to the car's control interface (likely through ROS2 Humble). This will enable DreamerV3 to directly control the F1Tenth car in simulated or real-world environments. This issue provides detailed guidance on integrating DreamerV3 with the F1Tenth platform, including code examples and explanations.

tl;dr - Figure out what the result of running DreamerV3 is, change that to our car actions(steering, speed)

## Current Issue

The current DreamerV3 implementation is likely configured with a more general action space and needs to be specifically adapted for the F1Tenth's two-dimensional continuous control (steering and speed). Directly using the current action space might lead to incorrect or unsafe control of the vehicle.  Furthermore, setting up and working with DreamerV3, especially in a ROS context, can be challenging.

## Expected Behavior

DreamerV3 should output two continuous control signals: steering and speed. These signals must:

1. **Be Within Valid Ranges:** The steering signal should be within the F1Tenth's physical steering limits (e.g., -1.0 to 1.0, where -1.0 is full left, 1.0 is full right, and 0.0 is straight). The speed signal should be within the car's speed limits (e.g., 0.0 to a maximum speed, such as 10.0 m/s). These ranges must be configurable.
2. **Directly Control the Car:** The steering and speed signals should be directly transmittable to the F1Tenth's control interface (likely through ROS messages).
3. **Be Sampled from a Distribution:** DreamerV3 should internally manage the exploration-exploitation trade-off. The actions should be sampled from a probability distribution (e.g., a Gaussian distribution) to allow for exploration during training.

## Code Implementation (Detailed Steps)

1. **Action Space Definition (DreamerV3 Configuration):**
    * **Locate the Action Space Definition:** Find where the action space is defined in your DreamerV3 implementation. This is usually within the agent's initialization code.
    * **Define Continuous Action Space:** Define a continuous action space with a dimension of 2.
    * **Specify Action Ranges:** Set the minimum and maximum values for steering and speed.  Make these configurable, ideally through the configuration file.

    ```python
    # Example (adapt to your DreamerV3 library and configuration method)
    action_dim = 2
    steering_min = -1.0
    steering_max = 1.0
    speed_min = 0.0
    speed_max = 10.0

    config = {
        # ... other config ...
        'action_dim': action_dim,
        'action_ranges': [(steering_min, steering_max), (speed_min, speed_max)],
        # ...
    }

    dreamer_v3_agent = DreamerV3(config)  # Initialize with config
    ```

2. **Action Sampling (Internal to DreamerV3):** DreamerV3 handles action sampling.  Ensure the agent samples from a suitable distribution (usually Gaussian). Adjust exploration parameters (e.g., `exploration_noise`, `temperature`) in the configuration file to control the exploration-exploitation balance.  A higher temperature or more noise leads to more exploration.

3. **Action Mapping (F1Tenth Control - ROS Integration):**
    * **ROS Setup:** Ensure your ROS environment is correctly configured and that you can communicate with the F1Tenth car.
    * **ROS Node:** Create a ROS node (or adapt an existing one) for communication.
    * **Subscribers:** Subscribe to the `LaserScan` topic (or the topic providing LiDAR data).
    * **Publishers:** Create a publisher for the `AckermannDriveStamped` topic (or the topic controlling the F1Tenth's motors).
    * **Callback (or Control Loop):**
        * **LiDAR Preprocessing:**  Call the `preprocess_lidar_data` function (see step 5) to process the raw `LaserScan` message.
        * **Get Action:** `steering, speed = dreamer_v3_agent.get_action(observation)`
        * **Create ROS Message:** Create an `AckermannDriveStamped` message and set `steering_angle` and `speed` using the values from DreamerV3.
        * **Publish Message:** `control_pub.publish(control_msg)`

    ```python
    import rospy
    from ackermann_msgs.msg import AckermannDriveStamped

    # ... (ROS node initialization) ...
    control_pub = rospy.Publisher('/f1tenth/drive', AckermannDriveStamped, queue_size=1)

    def scan_callback(scan_msg):
        observation = preprocess_lidar_data(scan_msg)
        steering, speed = dreamer_v3_agent.get_action(observation)

        control_msg = AckermannDriveStamped()
        control_msg.drive.steering_angle = steering
        control_msg.drive.speed = speed

        control_pub.publish(control_msg)

    # ... (ROS spin) ...
    ```

4. **LiDAR Data Preprocessing:**
    * **Implement `preprocess_lidar_data(scan_msg)`:**
    * **Convert to Cartesian Coordinates:** Convert the polar coordinates (range, angle) to Cartesian coordinates (x, y). This is *essential* for many RL algorithms, including DreamerV3.
    * **Noise Filtering:** Use a median filter or other suitable method to reduce noise.
    * **Data Scaling/Normalization:**  Scale or normalize the data to a suitable range (e.g., [-1, 1] or [0, 1]). This is *highly recommended* for improving training stability.
    * **Format Data:** Return the processed data as a NumPy array or PyTorch tensor.

    ```python
    import numpy as np
    import math

    def preprocess_lidar_data(scan_msg):
        ranges = np.array(scan_msg.ranges)
        angles = np.linspace(scan_msg.angle_min, scan_msg.angle_max, len(ranges))

        # Convert to Cartesian coordinates
        x = ranges * np.cos(angles)
        y = ranges * np.sin(angles)

        # Remove inf and nan values
        x = np.nan_to_num(x)
        y = np.nan_to_num(y)

        # Median filter (example - improve filter implementation as needed)
        filtered_x = x  # Replace with proper median filtering
        filtered_y = y  # Replace with proper median filtering

        # Data scaling/normalization (example)
        max_range = 10.0  # Replace with the maximum range of your LiDAR
        normalized_x = filtered_x / max_range
        normalized_y = filtered_y / max_range

        processed_data = np.stack([normalized_x, normalized_y], axis=-1)  # Shape (num_points, 2)
        return processed_data
    ```

5. **Training Loop Integration:** Integrate the LiDAR preprocessing and action mapping into the DreamerV3 training loop.  Replace any Gym-specific environment interaction code. This will involve passing the processed LiDAR data as the observation to the `dreamer_v3_agent.get_action` method and using the returned actions to control the F1Tenth.

6. **Testing and Validation:** Test in simulation first.  When moving to the real robot, be extremely cautious.  Start in a safe, open area.  Have a kill switch ready.  Incrementally increase speed and complexity.  Log data to analyze behavior.

## References

* **DreamerV3 Paper:**  [Mastering Diverse Games through World ### Models](https://arxiv.org/abs/2212.02923)
* **ROS Documentation:**  [ROS Tutorials](http://www.ros.org/tutorials/)
